docs/dynamism.md

```diff
@@ -0,0 +1,220 @@
+returns the indices of all elements that are `0` in a tensor value. The shape
+cannot be known without evaluating the data, but it can often be compiled using
+bounded dynamism, spending extra memory on the potential output tensor size.
+
+Many data-dependent dynamic ops can be modelled using bounded dynamism, where an
+upper bound on a tensor size is specified, and hardware generally will implement
+this via tensor padding. Today there is some support for data-dependent dynamism
+in PyTorch/XLA and TensorFlow, but JAX does not currently trace operations which
+lead to data dependent dynamism.
+
+[shape-poly]:https://jax.readthedocs.io/en/latest/export/shape_poly.html
+
+## Exporting programs with dynamic dimensions
+
+See our StableHLO tutorials for information on how to export programs with
+dynamic batch sizes or sequence lengths:
+
+- [JAX Tutorial > Export with Dynamic Batch Size][jax-export-dynamic]
+- [PyTorch/XLA Tutorial > Export with Dynamic Batch Size][pytorch-export-dynamic]
+
+[jax-export-dynamic]:https://openxla.org/stablehlo/tutorials/jax-export#export_with_dynamic_batch_size
+[pytorch-export-dynamic]:https://openxla.org/stablehlo/tutorials/pytorch-export#export_with_dynamic_batch_dimension
+
+## Compiler passes for refining dynamic programs
+
+### Remove dynamism pass pipeline
+
+There are a few useful passes for refining shapes, conveniently they are all
+bundled in a pass pipeline [`createStablehloRemoveDynamismPipeline`][remove-dynamism]:
+
+```c++
+void createStablehloRemoveDynamismPipeline(OpPassManager &pm,
+                                           TypeRange refinedTypes);
+```
+
+### Individual passes for refining dynamism
+
+Individually, the passes that tend to be useful for shape refinement are:
+
+- [`stablehlo-refine-arguments`][refine-arguments] to replace input arguments
+  with concrete tensor types.
+- [`stablehlo-refine-shapes`][refine-shapes] to propagate the new input argument
+  shape information throughout the entire program.
+- [`stablehlo-canonicalize-dynamism`][canonicalize-dynamism] to replace dynamic
+  ops with their static variants.
+
+See these passes generated documentation for up-to-date information and examples
+on their functionality.
+
+[remove-dynamism]:https://github.com/openxla/stablehlo/blob/ff13c96e56b73c62dcbb5b34b69f5ece9e71322f/stablehlo/transforms/Passes.h#L134
+[canonicalize-dynamism]:https://openxla.org/stablehlo/generated/stablehlo_passes#-stablehlo-canonicalize-dynamism
+[refine-arguments]:https://openxla.org/stablehlo/generated/stablehlo_passes#-stablehlo-refine-arguments
+[refine-shapes]:https://openxla.org/stablehlo/generated/stablehlo_passes#-stablehlo-refine-shapes
+
+## Example: How is dynamism useful, and how can I use it?
+
+Dynamism has lots of uses, here we'll mainly focus in on the common use case for
+Shape Polymorphism - creating a flexible exported model representation,
+generally used to represent dynamic batch size or sequence length.
+
+### Dynamic add_one model
```

Comment:
Same header name as the following one. Did you mean to say `Static add_one model`?

Reply:
good catch, fixed. "Static add_one model"